<!DOCTYPE html>
<html>
  <head> </head>

  <body>
    <main id="main-doc">
      <section class="main-section" id="Major_reasons_why_an_exception_Occurs">
        <header>Major reasons why an exception Occurs</header>
        <ul>
          <li>Invalid user input</li>
          <li>Device failure</li>
          <li>Loss of network connection</li>
          <li>Physical limitations (out of disk memory)</li>
          <li>Physical limitations (out of disk memory)</li>
          <li>Opening an unavailable file</li>
        </ul>
        <p>
          Errors represent irrecoverable conditions such as Java virtual machine
          (JVM) running out of memory, memory leaks, stack overflow errors,
          library incompatibility, infinite recursion, etc. Errors are usually
          beyond the control of the programmer, and we should not try to handle
          errors.
        </p>
        <p>
          Let us discuss the most important part which is the differences
          between Error and Exception that is as follows:
        </p>
        <ul>
          <li>
            Error: An Error indicates a serious problem that a reasonable
            application should not try to catch.
          </li>
          <li>
            Exception: Exception indicates conditions that a reasonable
            application might try to catch.
          </li>
        </ul>
      </section>
      <section class="main-section" id="Exception_Hierarchy">
        <header>Exception Hierarchy</header>
        <p>
          All exception and error types are subclasses of class Throwable, which
          is the base class of the hierarchy. One branch is headed by Exception.
          This class is used for exceptional conditions that user programs
          should catch. NullPointerException is an example of such an exception.
          Another branch, Error is used by the Java run-time system(JVM) to
          indicate errors having to do with the run-time environment
          itself(JRE). StackOverflowError is an example of such an error.
        </p>
      </section>
      <section class="main-section" id="Types_of_Exceptions">
        <header>Types of Exceptions</header>
        <p>
          Java defines several types of exceptions that relate to its various
          class libraries. Java also allows users to define their own
          exceptions.
        </p>
        <p>Exceptions can be categorized in two ways:</p>
        <ol>
          <li>
            Built-in Exceptions
            <ul>
              <li>Checked Exception</li>
              <li>Unchecked Exception</li>
            </ul>
          </li>
          <li>User-Defined Exceptions</li>
        </ol>
        <p>
          Let us discuss the above-defined listed exception that is as follows:
        </p>
        <ol type="A">
          <li>
            Built-in Exceptions:
            <p>
              Built-in exceptions are the exceptions that are available in Java
              libraries. These exceptions are suitable to explain certain error
              situations.
            </p>
            <ul>
              <li>
                Checked Exceptions: Checked exceptions are called compile-time
                exceptions because these exceptions are checked at compile-time
                by the compiler.
              </li>
              <li>
                Unchecked Exceptions: The unchecked exceptions are just opposite
                to the checked exceptions. The compiler will not check these
                exceptions at compile time. In simple words, if a program throws
                an unchecked exception, and even if we didn't handle or declare
                it, the program would not give a compilation error.
              </li>
            </ul>
          </li>
          <li>
            User-Defined Exceptions:
            <p>
              Sometimes, the built-in exceptions in Java are not able to
              describe a certain situation. In such cases, users can also create
              exceptions, which are called 'user-defined Exceptions'.
            </p>
            <p>The advantages of Exception Handling in Java are as follows:</p>
            <ol>
              <li>Provision to Complete Program Execution</li>
              <li>
                Easy Identification of Program Code and Error-Handling Code
              </li>
              <li>Propagation of Errors</li>
              <li>Meaningful Error Reporting</li>
              <li>Identifying Error Types</li>
            </ol>
            <p>Methods to print the Exception information:</p>
            <ol>
              <li>
                printStackTrace() - This method prints exception information in
                the format of Name of the exception: description of the
                exception, stack trace.
              </li>
              <li>
                toString() - This method prints exception information in the
                format of Name of the exception: description of the exception.
              </li>
              <li>
                getMessage() -This method prints only the description of the
                exception.
              </li>
            </ol>
          </li>
        </ol>
      </section>
      <section class="main-section" id="How_Does_JVM_handle_an_Exception?">
        <header>How Does JVM handle an Exception?</header>
        <p>
          Default Exception Handling: Whenever inside a method, if an exception
          has occurred, the method creates an Object known as an Exception
          Object and hands it off to the run-time system(JVM). The exception
          object contains the name and description of the exception and the
          current state of the program where the exception has occurred.
          Creating the Exception Object and handling it in the run-time system
          is called throwing an Exception. There might be a list of the methods
          that had been called to get to the method where an exception occurred.
          This ordered list of the methods is called Call Stack. Now the
          following procedure will happen.
        </p>
        <ul>
          <li>
            The run-time system searches the call stack to find the method that
            contains a block of code that can handle the occurred exception. The
            block of the code is called an Exception handler.
          </li>
          <li>
            The run-time system starts searching from the method in which the
            exception occurred, and proceeds through the call stack in the
            reverse order in which methods were called.
          </li>
          <li>
            If it finds an appropriate handler, then it passes the occurred
            exception to it. An appropriate handler means the type of the
            exception object thrown matches the type of the exception object it
            can handle.
          </li>
          <li>
            If the run-time system searches all the methods on the call stack
            and couldn't have found the appropriate handler, then the run-time
            system handover the Exception Object to the default exception
            handler, which is part of the run-time system. This handler prints
            the exception information in the following format and terminates the
            program abnormally.
          </li>
        </ul>
      </section>
      <section class="main-section" id="How_Programmer_Handles_an_Exception?">
        <header>How Programmer Handles an Exception?</header>
        <p>
          Customized Exception Handling: Java exception handling is managed via
          five keywords: try, catch, throw, throws, and finally. Briefly, here
          is how they work. Program statements that you think can raise
          exceptions are contained within a try block. If an exception occurs
          within the try block, it is thrown. Your code can catch this exception
          (using catch block) and handle it in some rational manner.
          System-generated exceptions are automatically thrown by the Java
          run-time system. To manually throw an exception, use the keyword
          throw. Any exception that is thrown out of a method must be specified
          as such by a throws clause. Any code that absolutely must be executed
          after a try block completes is put in a finally block.
        </p>
        <h2>Need for try-catch clause(Customized Exception Handling)</h2>
        <p>How to Use the try-catch Clause?</p>
        <p>
          Certain below key points are needed to be remembered that are as
          follows:
        </p>
        <ol>
          <li>
            In a method, there can be more than one statement that might throw
            an exception, So put all these statements within their own try block
            and provide a separate exception handler within their own catch
            block for each of them.
          </li>
          <li>
            If an exception occurs within the try block, that exception is
            handled by the exception handler associated with it. To associate
            the exception handler, we must put a catch block after it. There can
            be more than one exception handlers. Each catch block is an
            exception handler that handles the exception to the type indicated
            by its argument. The argument, ExceptionType declares the type of
            exception that it can handle and must be the name of the class that
            inherits from the Throwable class.
          </li>
          <li>
            For each try block, there can be zero or more catch blocks, but only
            one final block.
          </li>
          <li>
            The finally block is optional. It always gets executed whether an
            exception occurred in try block or not. If an exception occurs, then
            it will be executed after try and catch blocks. And if an exception
            does not occur, then it will be executed after the try block. The
            finally block in java is used to put important codes such as clean
            up code e.g., closing the file or closing the connection.
          </li>
        </ol>
      </section>
    </main>
  </body>
</html>

<script src="https://cdn.freecodecamp.org/testable-projects-fcc/v1/bundle.js"></script>
